monix-tail/shared/src/main/scala/monix/tail/internal/IterantIntersperse.scala

```diff
@@ -0,0 +1,114 @@
+import monix.execution.misc.NonFatal
+import monix.tail.Iterant
+import monix.tail.Iterant._
+import monix.tail.batches.BatchCursor
+
+
+private[tail] object IterantIntersperse {
+  def apply[F[_], A](source: Iterant[F, A], separator: A)(implicit F: Sync[F]): Iterant[F, A] = {
+    def processNonEmptyCursor(ref: NextCursor[F, A]): Iterant[F, A] = {
+      val NextCursor(cursor, rest, stop) = ref
+      val batchSize = cursor.recommendedBatchSize
+
+      if (batchSize <= 1) {
+        val item = cursor.next()
+        Next(item, F.pure(ref).map(loop(prepend = true)), stop)
+      } else {
+        var appends = 0
+        val maxAppends = batchSize / 2
+        val buffer = ArrayBuffer.empty[A]
+        var continue = true
+        while (continue && appends < maxAppends) {
+          buffer += cursor.next()
+          appends += 1
+          if (cursor.hasNext()) {
+            // only append separator if element is guaranteed to be not the last one
+            buffer += separator
+          } else {
+            continue = false
+          }
+        }
+        val batchCursor = BatchCursor.fromAnyArray[A](buffer.toArray[Any])
+        if (cursor.hasNext()) {
+          // ref now contains mutated cursor, continue with it
+          NextCursor(batchCursor, F.delay(loop(prepend = false)(ref)), stop)
+        } else {
+          NextCursor(batchCursor, rest.map(loop(prepend = true)), stop)
+        }
+      }
+    }
+
+    def loop(prepend: Boolean)(source: Iterant[F, A]): Iterant[F, A] = {
+      try source match {
+        case halt @ Halt(_) => halt
+
+        case Suspend(rest, stop) =>
+          Suspend(rest.map(loop(prepend)), stop)
+
+        case NextCursor(cursor, rest, stop) if !cursor.hasNext() =>
+          Suspend(rest.map(loop(prepend)), stop)
+
+        case NextBatch(batch, rest, stop) =>
+          val cursor = batch.cursor()
+          if (cursor.hasNext()) {
+            val processed =
+              processNonEmptyCursor(NextCursor(cursor, rest, stop))
+            if (prepend) {
+              Next(separator, F.pure(processed), stop)
+            } else {
+              processed
+            }
+          } else {
+            Suspend(rest.map(loop(prepend)), stop)
+          }
+
+        case _ if prepend => Next(
+          separator,
+          F.pure(source).map(loop(prepend = false)),
+          source.earlyStop
+        )
+
+        case ref @ NextCursor(_, _, _) =>
+          processNonEmptyCursor(ref)
+
+        case Next(item, rest, stop) =>
+          Next(item, rest.map(loop(prepend = true)), stop)
+
+        case last @ Last(_) => last
+      } catch {
+        case ex if NonFatal(ex) =>
+          val stop = source.earlyStop
+          Suspend(stop.map(_ => Halt(Some(ex))), stop)
+      }
+    }
+
+    source match {
+      case NextCursor(_, _, _) | NextBatch(_, _, _) =>
+        Suspend(F.delay(loop(prepend = false)(source)), source.earlyStop)
+      case _ => loop(prepend = false)(source)
```

Comment:
A minor matter of style / indentation — consistency is good as it makes it easier to read, probably because the arrangement is more predictable, so in a `match` statement if you have a branch with the expression on a new line above, it's better if you keep that for the rest of the branches, e.g...

```scala
source match {
  case NextCursor(_, _, _) | NextBatch(_, _, _) =>
    Suspend(F.delay(loop(prepend = false)(source)), source.earlyStop)
  case _ => 
    loop(prepend = false)(source) 
}
```